### ISSUE TITLE
Bug in Dumping a computational graph including LayerNormalization

### ISSUE THREAD

**Issue by F-Tag** [chainer/chainer#2215]
Added LayerNomalization to mnist example network.

```
class MLP(chainer.Chain):

    def __init__(self, n_units, n_out):
        super(MLP, self).__init__(
            # the size of the inputs to each layer will be inferred
            l1=L.Linear(None, n_units),  # n_in -> n_units
            ln1=L.LayerNormalization(),
            l2=L.Linear(None, n_units),  # n_units -> n_units
            ln2=L.LayerNormalization(),
            l3=L.Linear(None, n_out),  # n_units -> n_out
        )

    def __call__(self, x):
        h1 = F.relu(self.ln1(self.l1(x)))
        h2 = F.relu(self.ln2(self.l2(h1)))
        return self.l3(h2)
```

Dumped computational graph shows components of LayerNormalization.

![cg](https://cloud.githubusercontent.com/assets/25476669/22500329/21fe107a-e8a7-11e6-9631-c7f2c10e064d.png)

**Comment by delta2323** [comment]
Hi. TL;DR, it is difficult in current implementation that a link is expanded if `Link.__call__` is implemented with other links, functions, and variables.

It is Chainer functions and variables, not links that are appeared in dumped computational graphs. I think it would be natural to bundle Chainer functions and variables with the link to which these modules belong to . But currently, there is no way to tell Chainer functions and variables which link it belongs to automatically (i.e. without hints by the developer of the link). Also there is no way to tell a link which Chainer functions and variables belong to it.

I think analyzing such a hierarchical structure mentioned above without user's help would be as difficult as parsing Python scripts. On the other hand, it would be cumbersome for users to annotate links. It would be good to support "hierarchical" computational graphs. I am considering what is the appropriate compromising point.

**Comment by AkiTa24** [comment]
Thank you for reply.
I understood its difficulty.

I agree with "it would be natural to bundle Chainer functions and variables with the link to which these modules belong to."

I expect that it will be improved in the future.

**Comment by delta2323** [comment]
@niboshi Should we reopen this issue?

**Comment by niboshi** [comment]
@delta2323 I think so, as this is not solved yet.

**Comment by F-Tag** [comment]
This issue was solved by  layer normalization implemented as a function. (#2857).
So,  it would be nice to close this issue and make another issue ( about "hierarchical" computational graphs etc.) if necessary.

**Comment by delta2323** [comment]
Thank you for your comments. The problem of hierarchical computational graphs is addressed in #1424. Then we can close this issue with referencing the issue. Do you think it is reasonable? > @niboshi

**Comment by niboshi** [comment]
Yes, I agree with you.
